usr.sbin/ctm/ctm/ctm.1
.\"----------------------------------------------------------------------------
.\""THE BEER-WARE LICENSE" (Revision 42): 
.\"<joerg@freebsd.org> wrote this file.  As long as you retain this notice you
.\"can do whatever you want with this stuff. If we meet some day, and you think
.\"this stuff is worth it, you can buy me a beer in return.       Joerg Wunsch
.\"----------------------------------------------------------------------------
.\"
.\" This manual page is partially obtained from Poul-Hennings CTM README
.\" file.
.\"
.\" CTM and ctm(1) by <phk@login.dknet.dk>
.\"
.\" $Id: ctm.1,v 1.1 1996/10/30 17:32:57 graichen Exp $
.\"
.Dd Mar 25, 1995
.Os
.Dt CTM 1
.Sh NAME
.Nm ctm
.Nd source code mirror program
.Sh SYNOPSIS
.Nm ctm
.Op Fl cFpPquv
.Op Fl b Ar basedir
.Op Fl T Ar tmpdir
.Op Fl V Ar level
.Ar file Op ...
.Sh DESCRIPTION
.Nm Ctm
was originally
.Dq Cvs Through eMail ,
but now instead it seems more fitting to call it
.Dq Current Through eMail .

.Nm Ctm
is now meant to be the definitive way to make and apply a delta between
two versions of a directory tree.

There are two parts to this, making the delta and applying it.  These are two
entirely different things.

.Ss Usage

To apply a CTM delta, you pass it to the
.Nm ctm
command.  You can pass a CTM delta on stdin, or you can give the
filename as an argument.  If you do the latter, you make life a lot
easier for your self, since the program can accept gzip'ed files and
since it will not have to make a temporary copy of your file.  You can
specify multiple deltas at one time, they will be proccessed one at a
time.  Deltas that are already applied will be ignored.

The
.Nm ctm
command runs in a number of passes.  It will process the entire
input file in each pass, before commencing with the next pass.

Before working one a file
.Ar name
.Nm ctm
first checks for the existence of the file
.Ar name.ctm .
If this file exists,
.Nm ctm
works on it instead.

Pass 1 will validate that the input file is OK.  The syntax, the data
and the global MD5 checksum will be checked.  If any of these fail,
.Nm ctm
will never be able to do anything with the file, so it will simply
reject it.

Pass 2 will validate that the directory tree is in the state expected by
the CTM delta.  This is done by looking for files and directories which
should/should not exists and by checking the MD5 checksums of files.  

Pass 3 will actually apply the delta.

.Nm Ctm
will extract the file hierarchy below its working directory.  Absolute
filenames or filenames containing references through
.Sq \&.
and
.Sq \&.\&.
are explicitly prohibited as a security measure.

.Ss Options

.Bl -tag -width indent -compact

.It Fl b Ar basedir
Prepend the path
.Ar basedir
on every filename.

.It Fl c
Check it out, don't do anything.

.It Fl F
Force.

.It Fl p
Less paranoid.

.It Fl P
Paranoid.

.It Fl q
Tell us less.

.It Fl T Ar tmpdir
Put temporary files under
.Ar tmpdir .

.It Fl u
Set modification time of created and modified files to the CTM delta
creation time.

.It Fl v
Tell us more.

.It Fl V Ar level
Tell us more.
.Ar Level
is the level of verbosity.

.El

.Sh ENVIRONMENT
.Ev TMPDIR,
if set to a pathname, will cause ctm to use that pathname
as the location of temporary file.
See
.Xr tempnam 3,
for more details on this.
The same effect may be achieved with the
.Fl T
flag.

.Sh FILES

.Pa .ctm_status
contains the sequence number of the last CTM delta applied.  Changing
or removing this file will greatly confuse
.Nm ctm .

.Sh EXAMPLES

.Bd -literal

cd ~cvs
/usr/sbin/ctm ~ctm/cvs-*

.Ed

.Sh DIAGNOSTICS

Numerous messages, hopefully self-explaining.  The
.Dq noise level
can be adjusted with the
.Fl q
and
.Fl v
options.

.Sh SEE ALSO
.Xr ctm 5 ,
.Xr ctm_rmail 1

.Sh HISTORY

Initial trials ran during the FreeBSD 1.1.5, and many bugs and 
methods were hashed out.

The
.Nm ctm
command appeared in FreeBSD 2.1.

.Sh AUTHORS

The CTM system has been designed and implemented by
Poul-Henning Kamp
.Aq phk@FreeBSD.org .

Joerg Wunsch wrote this man-page.
.Aq joerg@FreeBSD.org .


usr.sbin/ctm/ctm/ctm.c
/*
 * ----------------------------------------------------------------------------
 * "THE BEER-WARE LICENSE" (Revision 42):
 * <phk@login.dknet.dk> wrote this file.  As long as you retain this notice you
 * can do whatever you want with this stuff. If we meet some day, and you think
 * this stuff is worth it, you can buy me a beer in return.   Poul-Henning Kamp
 * ----------------------------------------------------------------------------
 *
 * $Id: ctm.c,v 1.1 1996/10/30 17:32:58 graichen Exp $
 *
 * This is the client program of 'CTM'.  It will apply a CTM-patch to a
 * collection of files.
 *
 * Options we'd like to see:
 *
 * -a 			Attempt best effort.
 * -B <file>		Backup to tar-file.
 * -d <int>		Debug TBD.
 * -m <mail-addr>	Email me instead.
 * -r <name>		Reconstruct file.
 * -R <file>		Read list of files to reconstruct.
 *
 * Options we have:
 * -b <dir>		Base-dir
 * -c			Check it out, don't do anything.
 * -F      		Force
 * -p			Less paranoid.
 * -P			Paranoid.
 * -q 			Tell us less.
 * -T <tmpdir>.		Temporary files.
 * -u			Set all file modification times to the timestamp
 * -v 			Tell us more.
 * -V <level>		Tell us more level = number of -v
 *
 */

#define EXTERN /* */
#include "ctm.h"

#define CTM_STATUS ".ctm_status"

extern int Proc(char *, unsigned applied);

int
main(int argc, char **argv)
{
    int stat=0;
    int c;
    extern int optopt,optind;
    extern char * optarg;
    unsigned applied = 0;
    FILE *statfile;
    u_char * basedir;

    basedir = NULL;
    Verbose = 1;
    Paranoid = 1;
    SetTime = 0;
    setbuf(stderr,0);
    setbuf(stdout,0);

    while((c=getopt(argc,argv,"ab:B:cd:Fm:pPqr:R:T:uV:v")) != -1) {
	switch (c) {
	    case 'b': basedir = optarg;	break; /* Base Directory */
	    case 'c': CheckIt++;	break; /* Only check it */
	    case 'p': Paranoid--;	break; /* Less Paranoid */
	    case 'P': Paranoid++;	break; /* More Paranoid */
	    case 'q': Verbose--;	break; /* Quiet */
	    case 'v': Verbose++;	break; /* Verbose */
	    case 'T': TmpDir = optarg;	break;
	    case 'F': Force = 1;	break;
	    case 'u': SetTime++;	break; /* Set timestamp on files */
	    case 'V': sscanf(optarg,"%d", &c); /* Verbose */
		      Verbose += c;
		      break;
	    case ':':
		fprintf(stderr,"Option '%c' requires an argument.\n",optopt);
		stat++;
		break;
	    case '?':
		fprintf(stderr,"Option '%c' not supported.\n",optopt);
		stat++;
		break;
	    default:
		fprintf(stderr,"Option '%c' not yet implemented.\n",optopt);
		break;
	}
    }

    if(stat) {
	fprintf(stderr,"%d errors during option processing\n",stat);
	return Exit_Pilot;
    }
    stat = Exit_Done;
    argc -= optind;
    argv += optind;

    if (basedir == NULL) {
	Buffer = (u_char *)Malloc(BUFSIZ + strlen(SUBSUFF) +1);
	CatPtr = Buffer;
	*Buffer  = '\0';
    } else {
	Buffer = (u_char *)Malloc(strlen(basedir)+ BUFSIZ + strlen(SUBSUFF) +1);
	strcpy(Buffer, basedir);
	CatPtr = Buffer + strlen(basedir);
	if (CatPtr[-1] != '/') {
		strcat(Buffer, "/");
		CatPtr++;
	}
    }
    strcat(Buffer, CTM_STATUS);

    if((statfile = fopen(Buffer, "r")) == NULL)
	fprintf(stderr, "Warning: %s not found.\n", Buffer);
    else {
	fscanf(statfile, "%*s %u", &applied);
	fclose(statfile);
    }

    if(!argc)
	stat |= Proc("-", applied);

    while(argc-- && stat == Exit_Done) {
	stat |= Proc(*argv++, applied);
	stat &= ~Exit_Version;
    }

    if(stat == Exit_Done)
	stat = Exit_OK;

    if(Verbose)
	fprintf(stderr,"Exit(%d)\n",stat);
    return stat;
}

int
Proc(char *filename, unsigned applied)
{
    FILE *f;
    int i;
    char *p = strrchr(filename,'.');

    if(!strcmp(filename,"-")) {
	p = 0;
	f = stdin;
    } else if(p && (!strcmp(p,".gz") || !strcmp(p,".Z"))) {
	p = alloca(20 + strlen(filename));
	strcpy(p,"gunzip < ");
	strcat(p,filename);
	f = popen(p,"r");
	if(!f) { perror(p); return Exit_Garbage; }
    } else {
	p = 0;
	f = fopen(filename,"r");
    }
    if(!f) {
	perror(filename);
	return Exit_Garbage;
    }

    if(Verbose > 1)
	fprintf(stderr,"Working on <%s>\n",filename);

    Delete(FileName);
    FileName = String(filename);

    /* If we cannot seek, we're doomed, so copy to a tmp-file in that case */
    if(!p &&  -1 == fseek(f,0,SEEK_END)) {
	char *fn = tempnam(TmpDir,"CTMclient");
	FILE *f2 = fopen(fn,"w+");
	int i;

	if(!f2) {
	    perror(fn);
	    fclose(f);
	    return Exit_Broke;
	}
	unlink(fn);
	fprintf(stderr,"Writing tmp-file \"%s\"\n",fn);
	while(EOF != (i=getc(f)))
	    if(EOF == putc(i,f2)) {
		fclose(f2);
		return Exit_Broke;
	    }
	fclose(f);
	f = f2;
    }

    if(!p)
	rewind(f);

    if((i=Pass1(f, applied)))
	goto exit_and_close;

    if(!p) {
        rewind(f);
    } else {
	pclose(f);
	f = popen(p,"r");
	if(!f) { perror(p); return Exit_Broke; }
    }

    i=Pass2(f);

    if(!p) {
        rewind(f);
    } else {
	pclose(f);
	f = popen(p,"r");
	if(!f) { perror(p); return Exit_Broke; }
    }

    if(i) {
	if((!Force) || (i & ~Exit_Forcible))
	    goto exit_and_close;
    }

    if(CheckIt) {
        fprintf(stderr,"All checks out ok.\n");
	i = Exit_Done;
	goto exit_and_close;
    }

    i=Pass3(f);

exit_and_close:
    if(!p)
        fclose(f);
    else
	pclose(f);

    if(i)
	return i;

    fprintf(stderr,"All done ok\n");
    return Exit_Done;
}
